internal/controller/authenticator/jwtcachefiller/jwtcachefiller.go

```diff
@@ -210,16 +299,194 @@ func newJWTAuthenticator(spec *auth1alpha1.JWTAuthenticatorSpec) (*jwtAuthentica
 				},
 			},
 		},
-		KeySet:               coreosoidc.NewRemoteKeySet(ctx, providerJSON.JWKSURL),
+		KeySet:               coreosoidc.NewRemoteKeySet(ctx, jwksURL),
```

Comment:
I just had an idea. Curious what you think.

Before we call `oidc.New`, we could construct this remote key set and call its `VerifySignature` function and pass a hardcoded legal JWT that we know doesn't have the right signature. Then we could look at the error that it returns. If it returns `failed to verify id token signature` then we know that it successfully fetched the JWKS from the server and cached it internally for future use during real end-user authentication. If it returns an error that starts with `fetching keys` then we know it could not fetch the JWKS from the server. This would give us another validation for the user, and it would also pre-warm the authenticator's cache of JWKS. If it can't fetch the JWKS then it won't be able to authenticate end-users, so there's not really any point in loading it into the authenticator cache. The sync loop could return an error and retry in this case.

This does depend on our knowledge of the internal implementation of the remote key set library. We could use unit tests to make sure that we know if its error message behavior that we are relying upon ever changes during a future upgrade of that package. We couldn't be sure that the internal caching behavior didn't change on a future upgrade, but I would be very surprised if it ever changed because the JWKS fetching and caching is the whole point of the remote key set.

I don't think this is necessarily required for this PR, but might be a nice little enhancement. It would be an alternate solution to #771, but with the added advantage that the remote key set will refetch JWKS in the future as needed during the life of the authenticator.

Reply:
I like this idea, circling back when the other issues are all resolved.

Reply:
Latest commit takes a pass at implementing this.